### ISSUE TITLE
Building F* using F# on Windows without VS

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#159]
On Windows we can only build the F# variant of F* using Visual Studio.
Currently building the F# variant of F* from cygwin produces bad binary (see #132).
```
hritcu@W7-VBox /cygdrive/c/Users/hritcu/Desktop/FStar
$ ./bin/fstar.exe

Unexpected error; please file a bug report, ideally with a minimized version of the source program that triggered the error.
The type initializer for '<StartupCode$FSharp-PowerPack-Compatibility>.$Microsoft.FSharp.Compatibility.OCaml.Sys' threw an exception.
```

Antoine seemed to imply that it might be possible to do this right by passing flags like `--noframework` and `-r:FStar\bin\FSharp.PowerPack.dll` to `fsc.exe`. We should investigate this at some point.

**Comment by markulf** [comment]
Note that this also breaks "make -C src test" on Windows/

**Comment by catalin-hritcu** [comment]
Strangely enough, this seems to work on Nik's machine (but no other).

**Comment by markulf** [comment]
Investigated a bit further. It breaks if I compile FStar from the OCaml source, as that is faster than VS, but then run "make test". Is that expected?

**Comment by msprotz** [comment]
It mysteriously started working on my machine too after it had been failing for a while... and I'm completely incapable of figuring out why?!

**Comment by nikswamy** [comment]
I've since rebuilt my machines multiple times and this "make -C src" continues to work well for me.

**Comment by s-zanella** [comment]
I'm still getting an unsable binary when I build an F# binary using make rather than VS in Cygwin.

**Comment by nikswamy** [comment]
We suspect that building it once using VS enables future builds via Cygwin/make to succeed ... but not sure.

**Comment by mlr-msft** [comment]
`make -C src test` fails consistently on my Windows machine. 

furthermore, the use case where one checks out the git repo, build the OCaml snapshot (`make -C src/ocaml-output`), and attempt to run the regression suite (`make -C src regressions`) now fails consistently, from what i can tell.

if i build F* with Visual Studio, it produces `bin/tests.exe`. the regression suites run.

unfortunately, `bin/tests.exe` references `bin/fstar.exe` as a dependency. if i replace `bin/fstar.exe` with the OCaml version, this breaks `bin/tests.exe` and i can no longer run the regressions. if i don't build the F# version of F*, i don't get `bin/tests.exe`.

so, in short:

- `make -C src test` doesn't produce a working version of F* on Windows.
- the regressions can no longer be run for the OCaml snapshot, at least on Windows.

my workaround is to comment out the following line in `src/Makefile`, so i can run regression tests against the OCaml snapshot:
``` 
	$(BIN)/tests.exe
```

**Comment by nikswamy** [comment]
I think a pre-requisite for developing F\* is to be able to build the F#/F\* shared-fragment sources directly using fsc.

a. On a non-windows machine, this seems to work fine with mono.

b. On a windows machine, for most of us, a combination of a build using VS/make -C src does the job.

I don't understand what makes Michael's configuration so different that makes case b infeasible.

**Comment by msprotz** [comment]
Just to elaborate a little bit on top of Nik's comment:
* our belief is that by fiddling around and i) first building with VS then ii) building with the Makefile, then it is possible for everyone to attain a state where `make` succeeds reliably in building a working F* using `fsc` from the command-line
* the reason why sometimes `make` generates a buggy F* .NET binary has eluded us; neither Nik or I are .NET gurus; I did investigate a little bit, but didn't get anywhere and was content with the workaround
* fixing the present issue (`make` sometimes generate a faulty .NET F* binary) would certainly be appreciated and would make life easier for everyone
* Nik claims that by using mono on Windows, the present issue can be side-stepped.

**Comment by mlr-msft** [comment]
after examining the differences between how visual studio builds f* and how the gmake `Makefile` attempts to accomplish the same, i've succeeded in getting a working `tests.exe` on windows from the gmake, which runs successfully, so i hope it means that the `fstar.exe` is working as well.

i needed to do the following things in order to accomplish this:
- add the `--noframework` option to gmake's invocation of `fsc.exe`, along with `-r ...` options with absolute paths to every framework assembly referenced by the target (incl. `FSharp.Core.dll`).
- build the visual studio solution once, to ensure that all referenced assemblies are copied into the `bin` directory. 
- `make -C src clean`
- `make -C src`
- `bin/tests.exe`

here's what i suspect is happening:

when building with mono, there's generally one version of the framework installed at any one time. it's installed into a canonical location and there's no need to detect how the build system is configured. the gmake approach that's currently in use works fine in this situation.

on windows, however, there's potentially multiple frameworks and once the framework is detected and selected, the assemblies from this framework must be communicated to `fsc.exe` somehow. msbuild (and visual studio) appears to solve this through some discovery magic, explicitly referencing framework assemblies with full paths, and then copying all referenced assemblies into the same directory as the target so that they're available at runtime. 

i can think of two solutions to the problem described in this issue:

1. we can continue providing a gmake `Makefile` to build `fstar.exe` and `tests.exe` using `fsc.exe`. in order for this to function correctly and portably, we need to:

  a. keep it in sync with the visual studio project. i think this is worth mentioning because i've already discovered a divergence in the source file specifications.
  
  b. write (or obtain) logic to identify and select an appropriate framework on the builder's computer.
  
  c.  explicitly reference each assembly that `fstar.exe` and `tests.exe` references in `fsc.exe` invocations (using the search path option `--lib` didn't appear to work, unfortunately).
  
  d. copy each assembly into the `bin` directory.

2. we can use dispense with the gmake `Makefile` that builds f* with `fsc.exe` and rely upon msbuild (`xbuild.exe` in mono) instead. the following changes need to be made before this will work:

  a. `packages.config` needs to be patched in order to work with mono. this change needs to be regression tested against the windows build.
  
  b. automatic nuget restore appears to not work with msbuild. `nuget.exe restore` needs to be invoked before invoking msbuild.
  
  c. mark executables downloaded with nuget as executable.
  
  d. add support to the makefile to select the msbuild build configuration (debug vs. release).

currently, both approaches are at least partially implemented in (or integrated into) gmake. 

i recommend approach 2. i have concerns with approach 1 being brittle (specifically 1a & 1b) but i realize that there may be overriding needs for having a gmake-based build of f#-based f* that i'm unaware of.

**Comment by protz** [comment]
Hey,

@nikswamy and I had a quick chat about it, and we agree that option 2) seems like the best way to go. We would be happy to see the duplication of Makefile-based build and VS-based build go away in favor of a single way to build F* using F#. Furthermore, if we get rid of the Makefile way of doing things, then the mysterious failures where the generated F* executable doesn't run are going to go away, right? (Because xbuild/msbuild is going to do whatever voodoo is needed...?). In that case, that seems like a strong argument for going with 2.

We probably want "make" to default to "make fsharp-release", and have another "make fsharp-debug", so that none of our numerous scripts/habits need to change.

Thanks for the detective work!

**Comment by mlr-msft** [comment]
> Furthermore, if we get rid of the Makefile way of doing things, then the mysterious failures where the generated F* executable doesn't run are going to go away, right? (Because xbuild/msbuild is going to do whatever voodoo is needed...?)

right; that is my hypothesis. 

f* is complaining, at least on my machine, that it cannot find framework assemblies is depends upon during it's runtime linking stage. i anticipate that msbuild will copy the required assemblies into the `bin` directory, mirroring what visual studio does, which we already know produces good builds of f*.

> We probably want "make" to default to "make fsharp-release", and have another "make fsharp-debug", so that none of our numerous scripts/habits need to change.

that's useful to know. thanks.

> Thanks for the detective work!

happy to help out :smile:

**Comment by msprotz** [comment]
I've enabled building-via-msbuild in 88aa705ef22d7b03ac53d75a43f35f1899153a18; the default configuration is "release" (via the followup commit). I'm assuming that anyone who wants to debug will use Visual Studio's `.sln` file.

The remaining item is get rid of the custom Makefile entirely and call xbuild on OSX and Linux. This is on my radar and will probably happen, eventually, but in the meanwhile, anyone is welcome to take a stab at it.

**Comment by catalin-hritcu** [comment]
Given that the original problem of this issue if solved, should we close this and track the rest separately if needed?

**Comment by msprotz** [comment]
The `xbuild` part is done so we can finally close the issue. Thanks to @mlr-msft for suggesting the right fix!
